### DISCUSSION TITLE
Emergency stop Function for cfclient

### DISCUSSION THREAD

**Discussion by jonas-kupferschmid** [bitcraze/discussions#698] [Q&A]
Hello, I am still a beginner but want to create a series of youtube tutorials to help fellow beginners be able to get started with programming the Crazyflie. But I require some technical help with it. Thank you in advance for helping out!

I wanted to make an emergency stop when I press the Enter button on my laptop. I saw that in one of your tutorials "Crazyflie Logging and parameter framework tutorial" 8min 40s: https://www.youtube.com/watch?v=chWrNh73YBw&list=PLj9XMmQVSr-C-ZXeWWyZgeZ7C4g6A5nPj&index=9
you guys did something similar. However, the video is 9 years old and I saw that most of it is completely different now. I saw that there is a variable emergencyStop now. Can I also add a param group with it and change it or is it smarter to create a new Parameter? And how can I let cfclient listen for key presses to enable emergency stop? (pressing the button or changing in parameters page is a bit slow)

Thank you!

**Comment by krichardsson** [comment]
Hi!

This is an area that we are currently working on (an extended supervisor and arming functionality, see https://github.com/bitcraze/crazyflie-firmware/pull/1272). One of the features will actually be an emergency stop button in the client (I think it is almost done, but might not be part of the PR yet). Perhaps it does not make sense for you to work on the same feature, but I have a suggestion for a similar project.

What I suggest is to write a separate emergency stop application, essentially a window with a big red emergency stop button(-ish). This would be a small program that a user can run on a computer that broadcasts an emergency stop message to all Crazyflies using a second radio. The idea is to completely separate the emergency stop functionality from the application that controls the Crayzflie (like a script or Crazyswarm). In 2017 we built a physical emergency stop button (see https://www.bitcraze.io/2017/05/demo-for-icra-2017/) and what I have in mind is an application that is doing the same thing. 

You can find the code from 2017 in the repo that is linked in the blog post, most of it will probably still work :-) 
There are a few different ways to do emergency stop in the Crazyflie, but the one that is used in this case is the [emergency stop message](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/crtp_localization/#emergency-stop). It can be broadcast without a proper connection, which means that some other script can connect to the Crazyflie to control it. 
There is also a [Reset emergency stop timeout message](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/crtp_localization/#reset-emergency-stop-timeout) that is intended to make sure an emergency stop is triggered if the emergency stop button stops working.

What do you think, would this be an interesting project? I would be happy to point you in the right direction for where and how to implement it.
